firehydrant/terraform-provider-firehydrant
# Refactor r/service, d/service, and d/services

## Description

This PR does some minor refactoring to make the service resource and data sources consistent and to match some of the patterns and best practices we want to keep going forward. This should not have change the way these resources and data sources function.

## Testing plan

The goal of this testing plan is to make sure the service resources and data sources are still working the same as before. To do that, these steps have you test existing resources and new resources.

Pre-requisites:
- You'll need Go 1.16 and the latest Terraform installed.
- You'll need an instance of FireHydrant running locally or an organization in production to test against.
- You'll need a bot token for the organization you plan to test against.

#### Setup
1. Create a new directory called `providers` somewhere you can easily access (but not inside the same folder as your Terraform config).
1. Build the provider by checking out this branch and running `make build`
1. Move the executable created to the `providers` directory you created earlier.
   ```
   mv terraform-provider-firehydrant /PATH/TO/YOUR/DIRECTORY/providers
   ```
1. Add the following to your ~/.terraformrc file:
   ```
   provider_installation {

     # Use /PATH/TO/YOUR/DIRECTORY/providers/terraform-provider-firehydrant
     # as an overridden package directory for the firehydrant/firehydrant provider. 
     # This disables the version and checksum verifications for this provider and 
     # forces Terraform to look for the null provider plugin in the given directory.
     #dev_overrides {
     #  "firehydrant/firehydrant" = "/PATH/TO/YOUR/DIRECTORY/providers"
     #}

     # For all other providers, install them directly from their origin provider
     # registries as normal. If you omit this, Terraform will _only_ use
     # the dev_overrides block, and so no other providers will be available.
     direct {}
   }
   ```
1. Create another new directory (all Terraform commands will be run in this directory)
1.  [Download and save this file](https://github.com/firehydrant/terraform-provider-firehydrant/files/8119775/terraform-service-refactor-config.txt)

 as `main.tf` in your new directory. You'll be modifying it in the steps below. Make sure you replace the id in the service-by-id data source with the id of an existing service you have access to.
1. Run `terraform init`.

#### Upgrading - Existing services:
1. Run `terraform apply`. When prompted for an API key, provide your bot token. This should succeed and if you check in the UI, you should see the services you just created.
1. Edit your ~/.terraformrc file and remove the comments in front of the dev_overrides block
   ```
   dev_overrides {
     "firehydrant/firehydrant" = "/PATH/TO/YOUR/DIRECTORY/providers"
   }
   ```
1.  Run `terraform plan`. This time you should see a yellow warning block telling you that development overrides are in place. Your plan should show no changes. Go ahead and run `terraform apply -refresh-only` to pick up state updates from some of the other unreleased attributes.
1. Run `terraform destroy`. This should succeed and should remove any services you created.

#### Happy path - Create:
1. Run `terraform apply`. This should succeed and if you check in the UI, you should see the services you just created.

#### Happy path - Update:
1. Uncomment the alert_on_add and owner_id attributes for service5 and run `terraform apply`. This should succeed and should update these values in the UI.
1. Change service5's owner_id to this team2's ID:
   ```
   resource "firehydrant_service" "service5" {
     name         = "service5"
     description  = "description5"
     owner_id     = firehydrant_team.team2.id
     service_tier = 5
     labels = {
       test = "four"
     }
   }
   ```
1. Run `terraform apply`. This should show that the owner_id on service4 will change. It should succeed and you should see these changes reflected in the UI.
1. Remove the owner_id from service5 and run `terraform apply` again. This should succeed and should remove the owner from the UI.

#### Happy path - Destroy:
1. Run `terraform destroy`. This should succeed and should remove any services you created.

## Related links

- [Provider best practices](https://www.terraform.io/plugin/sdkv2/best-practices/detecting-drift#detecting-drift)

## PR readiness
